GVFS/GVFS.Service/RepoAutoMounter.cs

```diff
@@ -0,0 +1,150 @@
+            this.userSid = this.mountProcessManager.CurrentUser.Identity.User?.Value;
+        }
+
+        public void Start()
+        {
+            this.tracer.RelatedInfo("Starting auto mounter for session {0}", this.sessionId);
+
+            // Try mounting all the user's active repo straight away
+            this.tracer.RelatedInfo("Attempting to mount all known repos for user {0}", this.userSid);
+            this.MountAll();
+
+            // Start watching for changes to volume availability
+            this.volumeWatcher = GVFSPlatform.Instance.FileSystem.CreateVolumeStateWatcher();
+            this.volumeWatcher.VolumeStateChanged += this.OnVolumeStateChanged;
+            this.volumeWatcher.Start();
+        }
+
+        public void Stop()
+        {
+            this.tracer.RelatedInfo("Stopping auto mounter for session {0}", this.sessionId);
+
+            // Stop watching for changes to volume availability
+            if (this.volumeWatcher != null)
+            {
+                this.volumeWatcher.Stop();
+                this.volumeWatcher.VolumeStateChanged -= this.OnVolumeStateChanged;
+                this.volumeWatcher.Dispose();
+                this.volumeWatcher = null;
+            }
+        }
+
+        public void Dispose()
+        {
+            this.Stop();
+            this.mountProcessManager?.Dispose();
+        }
+
+        private void MountAll(string rootPath = null)
+        {
+            if (this.repoRegistry.TryGetActiveReposForUser(this.userSid, out var activeRepos, out string errorMessage))
+            {
+                foreach (RepoRegistration repo in activeRepos)
+                {
+                    if (rootPath == null || GVFSPlatform.Instance.FileSystem.IsPathUnderDirectory(rootPath, repo.EnlistmentRoot))
+                    {
+                        this.Mount(repo.EnlistmentRoot);
+                    }
+                }
+            }
+            else
+            {
+                this.tracer.RelatedError("Could not get repos to auto mount for user. Error: " + errorMessage);
+            }
+        }
+
+        private void Mount(string enlistmentRoot)
+        {
+            var metadata = new EventMetadata
+            {
+                ["EnlistmentRoot"] = enlistmentRoot
+            };
+
+            using (var activity = this.tracer.StartActivity("AutoMount", EventLevel.Informational, metadata))
+            {
+                string volumeRoot = GVFSPlatform.Instance.FileSystem.GetVolumeRoot(enlistmentRoot);
+                if (GVFSPlatform.Instance.FileSystem.IsVolumeAvailable(volumeRoot))
+                {
+                    // TODO #1043088: We need to respect the elevation level of the original mount
+                    if (this.mountProcessManager.StartMount(enlistmentRoot))
+                    {
+                        this.SendNotification("GVFS AutoMount", "The following GVFS repo is now mounted:\n{0}", enlistmentRoot);

```

Comment:
Aside: these toast notifications (existing code) from "GVFS.Service.UI" never actually seemed to show for me, even on released versions of VFS for Git.

Reply:
Yea they have been very problematic and inconsistent for us. I suspect something changed in the platform and we're no longer using their APIs correctly.